File: chess/java/src/main/java/com/example/chess/model/Board.java
package com.example.chess.model;

import com.example.chess.model.pieces.Bishop;
import com.example.chess.model.pieces.King;
import com.example.chess.model.pieces.Knight;
import com.example.chess.model.pieces.Pawn;
import com.example.chess.model.pieces.Piece;
import com.example.chess.model.pieces.Queen;
import com.example.chess.model.pieces.Rook;

import java.util.Optional;

/**
 * 8x8 board with piece storage and board-level helpers.
 *
 * <p>The board knows where pieces are and whether a sliding path is clear. It does not decide whose
 * turn it is; that orchestration belongs to {@code Game}. This separation keeps model rules focused.
 */
public class Board {
    public static final int SIZE = 8;

    private final Piece[][] grid = new Piece[SIZE][SIZE];

    /** Creates a standard chess starting position. */
    public Board() {
        setupInitialPosition();
    }

    private Board(boolean setup) {
        if (setup) {
            setupInitialPosition();
        }
    }

    /** Factory used by tests and demos that want to place only a few pieces. */
    public static Board empty() {
        return new Board(false);
    }

    public Optional<Piece> getPiece(Cell cell) {
        return Optional.ofNullable(grid[cell.row()][cell.col()]);
    }

    public void setPiece(Cell cell, Piece piece) {
        grid[cell.row()][cell.col()] = piece;
    }

    public Piece removePiece(Cell cell) {
        Piece piece = grid[cell.row()][cell.col()];
        grid[cell.row()][cell.col()] = null;
        return piece;
    }

    public void movePiece(Cell from, Cell to) {
        Piece piece = removePiece(from);
        setPiece(to, piece);
    }

    public boolean isEmpty(Cell cell) {
        return grid[cell.row()][cell.col()] == null;
    }

    public boolean hasOwnPiece(Cell cell, Color color) {
        return getPiece(cell).map(p -> p.getColor() == color).orElse(false);
    }

    public boolean hasEnemyPiece(Cell cell, Color color) {
        return getPiece(cell).map(p -> p.getColor() != color).orElse(false);
    }

    /** True when all squares strictly between {@code from} and {@code to} are empty. */
    public boolean isPathClear(Cell from, Cell to) {
        int rowStep = Integer.compare(to.row(), from.row());
        int colStep = Integer.compare(to.col(), from.col());
        int row = from.row() + rowStep;
        int col = from.col() + colStep;
        while (row != to.row() || col != to.col()) {
            if (grid[row][col] != null) {
                return false;
            }
            row += rowStep;
            col += colStep;
        }
        return true;
    }

    /** Locate a king, then ask every enemy piece polymorphically whether it attacks that square. */
    public boolean isInCheck(Color color) {
        Cell kingCell = findKing(color).orElseThrow(() -> new IllegalStateException("No king for " + color));
        return isSquareAttacked(kingCell, color.opposite());
    }

    public boolean isSquareAttacked(Cell target, Color byColor) {
        for (int r = 0; r < SIZE; r++) {
            for (int c = 0; c < SIZE; c++) {
                Piece piece = grid[r][c];
                if (piece != null && piece.getColor() == byColor
                        && piece.isValidMove(this, new Cell(r, c), target)) {
                    return true;
                }
            }
        }
        return false;
    }

    private Optional<Cell> findKing(Color color) {
        for (int r = 0; r < SIZE; r++) {
            for (int c = 0; c < SIZE; c++) {
                Piece piece = grid[r][c];
                if (piece instanceof King && piece.getColor() == color) {
                    return Optional.of(new Cell(r, c));
                }
            }
        }
        return Optional.empty();
    }

    private void setupInitialPosition() {
        setBackRank(0, Color.BLACK);
        setPawns(1, Color.BLACK);
        setPawns(6, Color.WHITE);
        setBackRank(7, Color.WHITE);
    }

    private void setPawns(int row, Color color) {
        for (int col = 0; col < SIZE; col++) {
            grid[row][col] = new Pawn(color);
        }
    }

    private void setBackRank(int row, Color color) {
        grid[row][0] = new Rook(color);
        grid[row][1] = new Knight(color);
        grid[row][2] = new Bishop(color);
        grid[row][3] = new Queen(color);
        grid[row][4] = new King(color);
        grid[row][5] = new Bishop(color);
        grid[row][6] = new Knight(color);
        grid[row][7] = new Rook(color);
    }
}

File: chess/python/chess/board.py
"""8x8 board storage and board-level chess helpers."""

from __future__ import annotations

from .models import Cell, Color
from .pieces import Bishop, King, Knight, Pawn, Piece, Queen, Rook


class Board:
    """Stores pieces and answers board questions such as path-clear and check.

    The board knows where pieces are. It does not decide whose turn it is; that orchestration belongs
    to ``Game``. This keeps the model rules focused and testable.
    """

    SIZE = 8

    def __init__(self, setup: bool = True) -> None:
        self._grid: list[list[Piece | None]] = [[None for _ in range(self.SIZE)] for _ in range(self.SIZE)]
        if setup:
            self._setup_initial_position()

    @classmethod
    def empty(cls) -> "Board":
        return cls(setup=False)

    def get_piece(self, cell: Cell) -> Piece | None:
        return self._grid[cell.row][cell.col]

    def set_piece(self, cell: Cell, piece: Piece | None) -> None:
        self._grid[cell.row][cell.col] = piece

    def remove_piece(self, cell: Cell) -> Piece | None:
        piece = self.get_piece(cell)
        self.set_piece(cell, None)
        return piece

    def move_piece(self, from_cell: Cell, to_cell: Cell) -> None:
        piece = self.remove_piece(from_cell)
        self.set_piece(to_cell, piece)

    def is_empty(self, cell: Cell) -> bool:
        return self.get_piece(cell) is None

    def has_own_piece(self, cell: Cell, color: Color) -> bool:
        piece = self.get_piece(cell)
        return piece is not None and piece.color is color

    def has_enemy_piece(self, cell: Cell, color: Color) -> bool:
        piece = self.get_piece(cell)
        return piece is not None and piece.color is not color

    def is_path_clear(self, from_cell: Cell, to_cell: Cell) -> bool:
        """True when all squares strictly between the two cells are empty."""
        row_step = (to_cell.row > from_cell.row) - (to_cell.row < from_cell.row)
        col_step = (to_cell.col > from_cell.col) - (to_cell.col < from_cell.col)
        row = from_cell.row + row_step
        col = from_cell.col + col_step
        while row != to_cell.row or col != to_cell.col:
            if self._grid[row][col] is not None:
                return False
            row += row_step
            col += col_step
        return True

    def is_in_check(self, color: Color) -> bool:
        king_cell = self._find_king(color)
        if king_cell is None:
            raise ValueError(f"No king for {color.value}")
        return self.is_square_attacked(king_cell, color.opposite)

    def is_square_attacked(self, target: Cell, by_color: Color) -> bool:
        for row in range(self.SIZE):
            for col in range(self.SIZE):
                piece = self._grid[row][col]
                if piece is not None and piece.color is by_color and piece.is_valid_move(self, Cell(row, col), target):
                    return True
        return False

    def _find_king(self, color: Color) -> Cell | None:
        for row in range(self.SIZE):
            for col in range(self.SIZE):
                piece = self._grid[row][col]
                if isinstance(piece, King) and piece.color is color:
                    return Cell(row, col)
        return None

    def _setup_initial_position(self) -> None:
        self._set_back_rank(0, Color.BLACK)
        self._set_pawns(1, Color.BLACK)
        self._set_pawns(6, Color.WHITE)
        self._set_back_rank(7, Color.WHITE)

    def _set_pawns(self, row: int, color: Color) -> None:
        for col in range(self.SIZE):
            self._grid[row][col] = Pawn(color)

    def _set_back_rank(self, row: int, color: Color) -> None:
        self._grid[row][0] = Rook(color)
        self._grid[row][1] = Knight(color)
        self._grid[row][2] = Bishop(color)
        self._grid[row][3] = Queen(color)
        self._grid[row][4] = King(color)
        self._grid[row][5] = Bishop(color)
        self._grid[row][6] = Knight(color)
        self._grid[row][7] = Rook(color)
